### PR TITLE
feat(metrics): add automatic media metrics and format

### PR DESCRIPTION
**Pull request by chrisadubois**
https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-122138

added MQA metrics for share

Screen Shot 2020-02-17 at 10 25 24 PM

Will be easy to add audio and video MQA, maybe 1-2 day effort.

works for FF and chromium

data verified in MQA backend

works for wireless share and non-wireless-share

does not break existing call flows

### CODE REVIEW

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 1]
```diff
@@ -2399,7 +2536,6 @@ export default class Meeting extends StatelessWebexPlugin {
   /**
    * Leave the current meeting
    * @param {Object} options An object of options
-   * @param {Boolean} options.keepLocalMedia Whether to tear down the local media connections or not.
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: What is the shape of the leave options object?

File: packages/node_modules/@webex/plugin-meetings/src/metrics/index.js [hunk 2]
```diff
@@ -116,7 +123,12 @@ class Metrics {
         buildType: 'prod',
         name: 'endpoint',
         networkType: 'unknown',
-        userAgent: this.getUserAgentInfo()
+        userAgent: this.getUserAgentInfo(),
+        clientInfo: {
+          clientType: options.clientType,
+          os: 'linux', // TODO: BUG AND DOESNT ACCEPT OSNAME ON CA -> bowser.osname || 'null'
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/metrics/index.js]: Is there a JIRA to track this TODO work?

File: packages/node_modules/@webex/plugin-meetings/src/metrics/index.js [hunk 3]
```diff
@@ -157,6 +169,36 @@ class Metrics {
     return payload;
   }
 
+  initMediaPayload(eventType, identifiers, options) {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/metrics/index.js]: jsdoc please

File: packages/node_modules/@webex/plugin-meetings/src/metrics/index.js [hunk 4]
```diff
@@ -157,6 +169,36 @@ class Metrics {
...
+    const payload = {
+      eventId: uuid.v4(),
+      version: 1,
+      origin: {
+        buildType: 'prod',
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/metrics/index.js]: Does this need to change when run on integration?

File: packages/node_modules/@webex/plugin-meetings/src/metrics/index.js [hunk 5]
```diff
@@ -157,6 +169,36 @@ class Metrics {
...
+      },
+      originTime: {
+        triggered: new Date().toISOString()
+      },
+      event: {
+        name: eventType,
+        canProceed: true,
+        identifiers,
+        intervals: [options.intervalData],
+        eventData: {webClientDomain: window.location.hostname}
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/metrics/index.js]: Do external users still want to use `webClientDomain`?

File: packages/node_modules/@webex/plugin-meetings/src/metrics/index.js [hunk 6]
```diff
@@ -299,22 +341,25 @@ class Metrics {
     return null;
   }
 
+  // TODO: we can add clientName to userAgent for CA to parse and add a config value
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/metrics/index.js]: Do we have a JIRA for this TODO?

File: packages/node_modules/@webex/plugin-meetings/src/metrics/mqa-processor.js [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    this.intervalNumber = 1;
+  }
+
+  /**
+   * @param {String} id
+   * @param {Array} interval - a slice of metrics history
+   * @returns {undefined}
+   */
+  process(id, interval) {
+    let rtcCandidatePair, rtcOutVideo, rtpOutVideo; // TODO: , rtcInVideo, rtpInVideo, rtcOutAudio, rtcInAudio, rtpInAudio, rtpOutAudio; // TODO:
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/metrics/mqa-processor.js]: Is there a JIRA for this TODO?

File: packages/node_modules/@webex/plugin-meetings/src/metrics/mqa-processor.js [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+      audioTransmit: [],
+      audioReceive: [],
+      videoTransmit: []
+    };
+    this.intervalNumber = 1;
+  }
+
+  /**
+   * @param {String} id
+   * @param {Array} interval - a slice of metrics history
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/metrics/mqa-processor.js]: Could you document the type of objects in the array?

Reference: https://jsdoc.app/tags-param.html#names-types-and-descriptions
```js
/**
 * Assign the project to a list of employees.
 * @param {Object[]} employees - The employees who are responsible for the project.
 * @param {string} employees[].name - The name of an employee.
 * @param {string} employees[].department - The employee's department.
 */
Project.prototype.assign = function(employees) {
    // ...
};
```

File: packages/node_modules/@webex/plugin-meetings/src/metrics/mqa-processor.js [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+        }
+        videoTransmit.push(vsTransmit);
+        break;
+      default:
+        break;
+    }
+    this.data.intervalMetadata = this.data.intervalMetadata || {...MQA_STATS.intervalMetadata};
+  }
+
+  getData() {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/metrics/mqa-processor.js]: JSDoc please

File: packages/node_modules/@webex/plugin-meetings/src/stats/metrics.js [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+  * @instance
+  * @type {Object}
+  * @memberof MediaMetrics
+  */
+
+/**
+ * @class MediaMetrics
+ * @private
+ */
+export default class MediaMetrics extends StatelessWebexPlugin {
```
Comment (adamweeks) [inline packages/node_modules/@webex/plugin-meetings/src/stats/metrics.js]: Does this need to extend `StatelessWebexPlugin`?

File: packages/node_modules/@webex/internal-plugin-metrics/src/call-diagnostic-events-batcher.js [hunk 11]
```diff
@@ -23,6 +23,7 @@ const CallDiagnosticEventsBatcher = Batcher.extend({
   prepareRequest(queue) {
     // Add sent timestamp
     queue.forEach((item) => {
+      item.eventPayload.originTime = item.eventPayload.originTime || {};
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-metrics/src/call-diagnostic-events-batcher.js]: Since this is a change to a separate plugin, it needs to be another commit please.

**Comment by adamweeks** [comment]
The "changed packages" test is reporting bad info, causing the integration tests to take too long and fail. Could you rebase on master please?

File: package-lock.json [hunk 12]
```diff
@@ -6929,7 +6929,8 @@
         },
         "ansi-regex": {
           "version": "2.1.1",
-          "bundled": true
```
[the remark this answers is not in the document]
Comment (chrisadubois) [reply package-lock.json]: ugh i thought i had removed this

File: packages/node_modules/@webex/plugin-meetings/src/config.js [hunk 13]
```diff
@@ -25,8 +25,20 @@ export default {
...
+    },
+    metrics: {
+      // change to your client name else data will be muddled
+      clientName: 'WEBEX_JS_SDK',
+      // TODO: for now this line has to be whitelisted, which is problematic for third party
+      clientType: 'TEAMS_CLIENT',
+      // send average values MQA in 60 second intervals
+      mqaMetricsInterval: 60000,
+      // send to cisco internal MQA data automatically
+      autoSendMQA: true
```
[the remark this answers is not in the document]
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/config.js]: I believe so, yes.

File: packages/node_modules/@webex/plugin-meetings/src/config.js [hunk 14]
```diff
@@ -25,8 +25,20 @@ export default {
...
     },
     stats: {
+      // 1 second intervals to collect stats data
       interval: 1000,
-      historyMax: 1800
+      // hold the last 2 minute of a calls data
+      historyMax: 120
+    },
+    metrics: {
+      // change to your client name else data will be muddled
```
[the remark this answers is not in the document]
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/config.js]: No

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 15]
```diff
@@ -570,45 +589,90 @@ export default class Meeting extends StatelessWebexPlugin {
...
+   * @param {String} options.trackingId
+   * @param {Object} options.locus
+   * @returns {undefined}
+   * @private
+   * @memberof Meeting
+   */
+  sendCallAnalyzerMetrics(options) {
+    const payload = this.getAnalyzerMetricsPrePayload(options);
+
+    this.webex.internal.metrics.submitCallDiagnosticEvents(payload);
```
[the remark this answers is not in the document]
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: We can but I don't think we need to, CA metrics shouldn't ever be blocking

File: packages/node_modules/@webex/plugin-meetings/src/meeting/index.js [hunk 16]
```diff
@@ -570,45 +589,90 @@ export default class Meeting extends StatelessWebexPlugin {
...
+   * @returns {undefined}
+   * @private
+   * @memberof Meeting
+   */
+  sendMediaQualityAnalyzerMetrics(options) {
+    options.type = MQA_STATS.CA_TYPE;
+    options.clientType = this.config.metrics.clientType;
+    const payload = this.getAnalyzerMetricsPrePayload(options);
 
     this.webex.internal.metrics.submitCallDiagnosticEvents(payload);
```
[the remark this answers is not in the document]
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/meeting/index.js]: same as above

File: packages/node_modules/@webex/plugin-meetings/src/stats/metrics.js (hunk 10 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/stats/metrics.js]: it did to access the config, though we could also use the StaticConfig. I am not sure what direction the SDK plugin arch is headed

File: packages/node_modules/@webex/plugin-meetings/src/metrics/index.js (hunk 4 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/metrics/index.js]: No, it only accepts prod as of now

File: packages/node_modules/@webex/plugin-meetings/src/metrics/index.js (hunk 5 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/metrics/index.js]: Yeah, it's how we differentiate it on our dashboard, we don't want to change the key

File: packages/node_modules/@webex/plugin-meetings/src/config.js [hunk 17]
```diff
@@ -25,8 +25,20 @@ export default {
...
+      clientName: 'WEBEX_JS_SDK',
+      // TODO: for now this line has to be whitelisted, which is problematic for third party
+      clientType: 'TEAMS_CLIENT',
```
[the remark this answers is not in the document]
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/config.js]: https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-122143

File: packages/node_modules/@webex/plugin-meetings/src/constants.js [hunk 18]
```diff
@@ -821,3 +860,77 @@ export const STATS = {
...
 };
 
+export const MQA_STATS = {
+  MQA_SIZE: 120, // MQA is done on 60 second intervals by server def, add a buffer for missed events
+  CA_TYPE: 'MQA',
+  DATA_PLACEMENTS: [STATS.AUDIO_SENDER_ID, STATS.AUDIO_RECEIVER_ID, STATS.VIDEO_SENDER_ID, STATS.VIDEO_RECEIVER_ID, STATS.SHARE_SENDER_ID, STATS.SHARE_RECEIVER_ID],
+  DEFAULT_SHARE_SENDER_STATS: {
+    common: {
+      common: {
+        direction: 'sendrecv', // TODO: parse from SDP and save globally
```
[the remark this answers is not in the document]
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/constants.js]: https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-122143

File: packages/node_modules/@webex/plugin-meetings/src/constants.js [hunk 19]
```diff
@@ -821,3 +860,77 @@ export const STATS = {
...
+      rtpBitrate: -1, // unavailable
+      rtpPackets: -1,
+      stunBitrate: -1, // unavailable
+      stunPackets: -1, // unavailable
+      transportType: 'UDP' // TODO: parse the transport type from the SDP and save globally
+    },
+    streams: [
+      {
+        common: {
+          codec: 'H264', // TODO: parse the codec from the SDP and save globally
```
[the remark this answers is not in the document]
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/constants.js]: https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-122143

File: packages/node_modules/@webex/plugin-meetings/src/constants.js [hunk 20]
```diff
@@ -821,3 +860,77 @@ export const STATS = {
...
+      remoteLossRate: -1,
+      remoteReceiveRate: -1, // unavailable
+      roundTripTime: -1,
+      rtcpBitrate: -1, // unavailable
+      rtcpPackets: -1, // unavailable
+      rtpBitrate: -1, // unavailable
+      rtpPackets: -1,
+      stunBitrate: -1, // unavailable
+      stunPackets: -1, // unavailable
+      transportType: 'UDP' // TODO: parse the transport type from the SDP and save globally
```
[the remark this answers is not in the document]
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/constants.js]: https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-122143

File: packages/node_modules/@webex/plugin-meetings/src/constants.js [hunk 21]
```diff
@@ -821,3 +860,77 @@ export const STATS = {
...
+          codec: 'H264', // TODO: parse the codec from the SDP and save globally
+          duplicateSsci: -1, // unavailable
+          requestedBitrate: -1, // unavailable
+          requestedFrames: -1, // unavailable
+          rtpPackets: -1,
+          ssci: -1, // unavailable
+          transmittedBitrate: -1,
+          transmittedFrameRate: -1
+        },
+        h264CodecProfile: 'BP', // TODO: parse the profile level from h264 in the SDP and save globally
```
[the remark this answers is not in the document]
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/constants.js]: https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-122143

File: packages/node_modules/@webex/plugin-meetings/src/constants.js [hunk 22]
```diff
@@ -821,3 +860,77 @@ export const STATS = {
...
+      cpuBitWidth: 0,
+      mainProcessMaximumMemoryBytes: 0,
+      osBitWidth: 0,
+      processAverageMemoryUsage: 0,
+      processMaximumMemoryBytes: 0,
+      processMaximumMemoryUsage: 0,
+      systemAverageMemoryUsage: 0,
+      systemMaximumMemoryUsage: 0
+    },
+    peerReflexiveIP: 'NULL', // TODO: save after ice trickling completes and use as a global variable
```
[the remark this answers is not in the document]
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/constants.js]: https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-122143

File: packages/node_modules/@webex/plugin-meetings/src/metrics/index.js (hunk 2 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/metrics/index.js]: https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-122143

File: packages/node_modules/@webex/plugin-meetings/src/metrics/index.js (hunk 6 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/metrics/index.js]: https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-122143

File: packages/node_modules/@webex/plugin-meetings/src/metrics/mqa-processor.js (hunk 7 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/metrics/mqa-processor.js]: https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-122143

File: packages/node_modules/@webex/internal-plugin-metrics/src/call-diagnostic-events-batcher.js (hunk 11 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/internal-plugin-metrics/src/call-diagnostic-events-batcher.js]: done: https://github.com/webex/webex-js-sdk/pull/1527

**Comment by chrisadubois** [comment]
@adamweeks i have addressed all the comments, created a jira, and opened a separate PR, please re-review, ty

File: packages/node_modules/@webex/plugin-meetings/src/config.js [hunk 23]
```diff
@@ -25,8 +25,22 @@ export default {
...
+      // change to your client name else data will be muddled
+      // you do not need a specific format, and you do not need to register it
+      clientName: 'WEBEX_JS_SDK',
+      // TODO: for now this line has to be whitelisted, which is problematic for third party
+      clientType: 'TEAMS_CLIENT',
+      // send average values MQA in 60 second intervals
+      mqaMetricsInterval: 60000,
+      // send to cisco internal MQA data automatically
+      // we already send CA
+      autoSendMQA: true
```
Comment (andrewholsted) [inline packages/node_modules/@webex/plugin-meetings/src/config.js]: I think this should default to false? @adamweeks what's the stance on us collecting data by default from third party devs? Do we do this anywhere else?

File: packages/node_modules/@webex/plugin-meetings/src/meeting/util.js [hunk 24]
```diff
@@ -154,6 +154,7 @@ MeetingUtil.cleanUp = (meeting) => meeting.closeLocalStream()
     meeting.unsetLocalShareTrack();
     meeting.unsetRemoteStream();
     meeting.unsetPeerConnections();
+    meeting.cleanMQAInterval();
```
Comment (andrewholsted) [inline packages/node_modules/@webex/plugin-meetings/src/meeting/util.js]: Nice

File: packages/node_modules/@webex/plugin-meetings/src/metrics/mqa-processor.js [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+        break;
+      case STATS.VIDEO_RECEIVER_ID:
+        videoReceive.push();
+        break;
+      case STATS.SHARE_SENDER_ID:
+        if (sumValue) {
+          rtcOutVideo = sumValue.rtcOutVideo;
+          rtpOutVideo = sumValue.rtpOutVideo;
+        }
+        vsTransmit = {...MQA_STATS.DEFAULT_SHARE_SENDER_STATS};
```
Comment (andrewholsted) [inline packages/node_modules/@webex/plugin-meetings/src/metrics/mqa-processor.js]: does `vsTransmit` need to be declared outside the switch?

File: packages/node_modules/@webex/plugin-meetings/src/config.js (hunk 23 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/config.js]: We send CA metrics by default, this was discussed with Matt months ago.

File: packages/node_modules/@webex/plugin-meetings/src/stats/metrics.js (hunk 10 above)
Comment (adamweeks) [reply packages/node_modules/@webex/plugin-meetings/src/stats/metrics.js]: Yes, please do not use the plugin class except for the top level plugin. We are going to need to refactor the meetings plugin to not subclass the plugin throughout their classes.

File: packages/node_modules/@webex/plugin-meetings/src/metrics/mqa-processor.js (hunk 25 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/metrics/mqa-processor.js]: eslint :(

File: packages/node_modules/@webex/plugin-meetings/src/config.js (hunk 23 above)
Comment (adamweeks) [reply packages/node_modules/@webex/plugin-meetings/src/config.js]: I agree, it should default to false. Third parties should opt-in to data collection.

**Comment by adamweeks** [comment]
Please move the test-helper-mock-webex change into its own commit. Refer to the [scope](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#scope) section of the contributing guide for more info.

File: packages/node_modules/@webex/plugin-meetings/src/stats/metrics.js (hunk 10 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/stats/metrics.js]: Sure, can change it.

File: packages/node_modules/@webex/plugin-meetings/src/config.js (hunk 23 above)
Comment (chrisadubois) [reply packages/node_modules/@webex/plugin-meetings/src/config.js]: Changing this value won't change anything now since code exists that will always send to CA. This will need to be addressed as tech debt going forward. Also it was signed off on previously by multiple levels so I want to make sure we have the time to discuss as a whole group. Changing that is out of scope in this PR.

Comment (adamweeks) [reply packages/node_modules/@webex/plugin-meetings/src/config.js]: img

File: packages/node_modules/@webex/plugin-meetings/src/metrics/mqa-processor.js (hunk 25 above)
Comment (adamweeks) [reply packages/node_modules/@webex/plugin-meetings/src/metrics/mqa-processor.js]: You can wrap the case into a block to be able to `const vsTransmit`
